pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,193 @@
+	Prefix = "grpc://"
+
+	// Connection retry constants
+	// Backoff period is calculated with backOffBase ^ #retry-iteration. However, when backoffLimit is reached, fallback
+	// to constantBackoffDelay
+	backoffLimit         = 3
+	backOffBase          = 4
+	constantBackoffDelay = 60
+)
+
+type Sync struct {
+	Target     string
+	ProviderID string
+	Logger     *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	options := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// initial dial and connection. Failure here must result in a startup failure
+	dial, err := grpc.DialContext(ctx, g.Target, options...)
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing grpc connection: %s", err.Error()))
+		return err
+	}
+
+	serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+	syncClient, err := serviceClient.SyncFlags(context.Background(), &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error calling streaming operation: %s", err.Error()))
+		return err
+	}
+
+	// initial stream listening
+	err = g.streamListener(ctx, syncClient, dataSync)
+	if err != nil {
+		g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+	}
+
+	// retry connection establishment
+	for {
+		g.Logger.Warn(fmt.Sprintf("Connection re-establishment attempt in-progress for grpc target: %s", g.Target))
+
+		syncClient = g.connectWithRetry(ctx, options...)
+		err = g.streamListener(ctx, syncClient, dataSync)
+		if err != nil {
+			g.Logger.Warn(fmt.Sprintf("Error with stream listener: %s", err.Error()))
+			continue
+		}
+	}
+}
+
+// connectWithRetry is a helper to perform exponential backoff till provided configurations and then retry connection
+// periodically till a successful connection is established
+func (g *Sync) connectWithRetry(
+	ctx context.Context, options ...grpc.DialOption,
+) syncv1grpc.FlagSyncService_SyncFlagsClient {
+	var iteration int
+
+	for {
+		var sleep time.Duration
+		if iteration >= backoffLimit {
+			sleep = constantBackoffDelay
+		} else {
+			iteration++
+			sleep = time.Duration(math.Pow(backOffBase, float64(iteration)))
+		}
+
+		time.Sleep(sleep * time.Second)
+
+		dial, err := grpc.DialContext(ctx, g.Target, options...)
+		if err != nil {
+			g.Logger.Debug(fmt.Sprintf("Error dialing target: %s", err.Error()))
+			continue
+		}
+
+		serviceClient := syncv1grpc.NewFlagSyncServiceClient(dial)
+		syncClient, err := serviceClient.SyncFlags(context.Background(), &v1.SyncFlagsRequest{ProviderId: g.ProviderID})
+		if err != nil {
+			g.Logger.Debug(fmt.Sprintf("Error openning service client: %s", err.Error()))
+			continue
+		}
+
+		g.Logger.Info(fmt.Sprintf("Connection re-established with grpc target: %s", g.Target))
+		return syncClient
+	}
+}
+
+// streamListener wraps the grpc listening on provided stream and push updates through dataSync channel
+func (g *Sync) streamListener(
```

Comment:
It looks like this function can actually be simplified to:
```
func (g *Sync) streamListener(
	stream syncv1grpc.FlagSyncService_SyncFlagsClient, dataSync chan<- sync.DataSync,
) error {
		err := g.handleFlagSync(stream, dataSync)
	if err == io.EOF {
		g.Logger.Info("Stream closed by the server")
	}
	return err
}
```
The `handleFlagSync` doesn't need to be run in a goroutine, it will block until the parent context closes and it either spits out an EOF, or another error.
Another option is to directly invoke `handleFlagSync` in the `Sync` method, and check the error / log the error there instead, allowing the `handleFlagSync` to block until it is shutdown by the parent context closing.

Reply:
Yes, correct. We can skip the `streamListener`. 

My initial implression was `stream.Recv()` to be a blocking call, without handling parent context. But I was wrong. Underlying stream is context aware per documentation [1] and it respect the context cancellations. 


[1] - https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream

Reply:
Updated with the latest commit